enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,87 @@
+---
+title: image-content-source-policy-awareness
+authors:
+  - "@sallyom"
+reviewers:
+  - "@smarterclayton"
+  - "@soltysh"
+  - "@wking"
+approvers:
+  - "@smarterclayton"
+creation-date: 2020-05-19
+last-updated: 2020-05-19
+status: provisional
+---
+
+# ImageContentSourcePolicy (ICSP) Awareness
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+There have been several bugs opened around the experience of a 
+user in a disconnected environment using `oc adm release` commands.  If
+using a mirrored image and the mirrored source registry is connected, 
+the following commands do not succeed when in a disconnected environment:
+
+```console
+$ oc adm release extract --tools aprivateregistry/repo/name:tag
+$ oc adm release mirror aprivateregistry/repo/name:tag --to someregistry/repo/name
+$ oc adm must-gather
+```
+
+This is because the mirrored image tags (the individual component images from a payload)
+retain references to the mirrored registry, usually something like 
+`quay.io/openshift-release-dev/ocp-v4.0-art-dev`.  
+
+There needs to be logic in `oc` to look for `ImageContentSourcePolicy` either from a file or from a cluster.
+`oc` should look for `ICSP` in the cluster/current context if connected, and if not, from a file in an expected location,
+probably from the current directory.  `oc` should gather information about RepositoryDigestMirrors from ICSP and use that
+when extracting or mirroring images, or when running must-gather.
+
+Current bugs regarding this Issue:   
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823839
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823143 and also for 4.3, 4.5, 4.6
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+* In what order should `oc` gather info about an image?  Need to order the following:
+    * ICSP from file
+    * ICSP from cluster
+    * image reference from user-given image
+
+* Should oc adm release mirror write an ICSP file to the current directory rather than a print to stdout?
+* Should there be a flag to direct oc to re-order the above 3 choices, to tell oc which information source to use?
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
```

Comment:
`oc` seems to use hard-coded image references, for example `oc adm must-gather` just tries to pull `quay.io/openshift/origin-must-gather:latest`. How will `oc` use ICSP to find right image? `oc`'s going to have to know the pullspec for the release and then do whatever `oc adm release info --image-for=must-gather` would do?

Reply:
Must gather should use the tag in the openshift image stream (it should be like debug, which tries to look up `openshift/tools:latest` and then falls back to a hardcoded constant).

Reply:
That makes sense but since must-gather is in the release image, can't that be added as a fallback option?

We don't deploy the samples operator on baremetal by default, and even if we did and it failed, I'd still expect must-gather to work in the disconnected case.

Reply:
For disconnected `oc adm must-gather` I believe you have to pass the 
`oc adm must-gather --image=disco-repo/release:4.5.0-0.ci-2020-05-26-081827-must-gather` flag, otherwise it will try to launch a pod using the imagestream from -n openshift and that still has reference to the quay.io registry from which the release image was mirrored.  When that fails must-gather will then try the hard-coded origin-must-gather:latest

Reply:
Right, but as an end user I'd just expect `oc adm must-gather` to work in a disconnected environment. Specifying the image isn't very intuitive.... we do it like this: https://github.com/openshift-metal3/dev-scripts/blob/master/must_gather.sh#L19-L20

Reply:
i see, ok

Reply:
@stbenjam must-gather will do the following:
1. resolve images passed through --image flag
2. resolve images streams passed through --image-stream flag
3. resolve must-gather:latest IStag from openshift ns
all of the above built the set of plugins/images that are being run during must-gather.
see https://github.com/openshift/oc/blob/f30826ef04e45eea1745276dee8e693a08e29dba/pkg/cli/admin/mustgather/mustgather.go#L141-L158

Reply:
@stbenjam similar logic applies to `oc debug` it will look for tools:latest in openshift ns, and fallback to registry only if none is found.